### ISSUE TITLE
Apple Silicon Tech Preview RC2 broke as a Vagrant provider

### ISSUE THREAD

**Issue by rofrano** [docker/for-mac#5527]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: EA5613D3-8F41-4508-8615-34E209F88E02/20210331014351

### Expected behavior

I have been using Docker Tech Preview as a provider for Vagrant because solutions like VirtualBox do not work on Apple Silicon. I expect the Docker container to work as a provider for vagrant with RC2 just as it did with Tech Preview 7 as follows: 

```shell
git clone https://github.com/nyu-devops/lab-flask-rest.git
cd lab-flask-rest 
vagrant up --provider=docker
```

This worked perfectly with Docker Apple Silicon Preview 7. It broke when I upgraded to Release Candidate 2. :(

### Actual behavior

Docker causes a failure when vagrant tries to update the SSH keys in the Docker container. The output of the failure is shown in the section below on [Steps to reproduce the behavior](#steps-to-reproduce-the-behavior):

### Information

  - macOS Version: 11.2.3
  - Docker Desktop Version: 3.3.0 (62345) RC2
  - The problem is completely reproducible. Several other people have confirmed the same problem.
  - The problem is new when upgrading to RC2, it worked in Preview 7

### Steps to reproduce the behavior

To reproduce, execute the following steps on a Mac with Apple Silicon:

  1. Install Vagrant 
  2. Clone the repo https://github.com/nyu-devops/lab-flask-rest.git
  3. Change into the `lab-flask-rest` directory
  4. Bring up vagrant with `vagrant up --provider=docker`

Here are the actual commands:

Install Vagrant with [Homebrew](https://brew.sh)

```shell
brew install vagrant
```

Clone the repo and bring the container up with Vagrant:

```shell
git clone https://github.com/nyu-devops/lab-flask-rest.git
cd lab-flask-rest 
vagrant up --provider=docker
```

This is the error output: 

```
Bringing machine 'default' up with 'docker' provider...
==> default: Creating and configuring docker networks...
==> default: Creating the container...
    default:   Name: lab-flask-rest_default_1617154699
    default:  Image: rofrano/vagrant-provider:ubuntu
    default: Volume: /Users/rofrano/DevOps/lab-flask-rest:/vagrant
    default:   Port: 127.0.0.1:5000:5000
    default:   Port: 127.0.0.1:2222:22
    default:  
    default: Container created: b2719c3905c61841
==> default: Enabling network interfaces...
==> default: Starting container...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default: 
    default: Inserting generated public key within guest...
Traceback (most recent call last):
	35: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/lib/vagrant/action/builtin/wait_for_communicator.rb:16:in `block in call'
	34: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:63:in `wait_for_ready'
	33: from /opt/vagrant/embedded/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
	32: from /opt/vagrant/embedded/lib/ruby/2.6.0/timeout.rb:33:in `catch'
	31: from /opt/vagrant/embedded/lib/ruby/2.6.0/timeout.rb:33:in `catch'
	30: from /opt/vagrant/embedded/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
	29: from /opt/vagrant/embedded/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
	28: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:88:in `block in wait_for_ready'
	27: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:192:in `ready?'
	26: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/lib/vagrant/guest.rb:43:in `capability'
	25: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/lib/vagrant/capability_host.rb:111:in `capability'
	24: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/lib/vagrant/capability_host.rb:111:in `call'
	23: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/guests/linux/cap/public_key.rb:14:in `insert_public_key'
	22: from /opt/vagrant/embedded/lib/ruby/2.6.0/tempfile.rb:295:in `open'
	21: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/guests/linux/cap/public_key.rb:19:in `block in insert_public_key'
	20: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:307:in `upload'
	19: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:746:in `scp_connect'
	18: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:392:in `connect'
	17: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:748:in `block in scp_connect'
	16: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:338:in `block in upload'
	15: from /opt/vagrant/embedded/gems/2.2.14/gems/vagrant-2.2.14/plugins/communicators/ssh/communicator.rb:332:in `block (2 levels) in upload'
	14: from /opt/vagrant/embedded/gems/2.2.14/gems/net-scp-1.2.1/lib/net/scp.rb:284:in `upload!'
	13: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/channel.rb:272:in `wait'
	12: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:180:in `loop'
	11: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:180:in `loop'
	10: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:180:in `block in loop'
	 9: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:227:in `process'
	 8: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/event_loop.rb:28:in `process'
	 7: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/event_loop.rb:100:in `ev_preprocess'
	 6: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/event_loop.rb:100:in `each'
	 5: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:248:in `ev_preprocess'
	 4: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:548:in `dispatch_incoming_packets'
	 3: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:685:in `channel_close'
	 2: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/session.rb:572:in `channel_closed'
	 1: from /opt/vagrant/embedded/gems/2.2.14/gems/net-ssh-6.2.0.rc1/lib/net/ssh/connection/channel.rb:611:in `do_close'
/opt/vagrant/embedded/gems/2.2.14/gems/net-scp-1.2.1/lib/net/scp.rb:365:in `block (3 levels) in start_command': SCP did not finish successfully (1):  (Net::SCP::Error)
```

For a more detailed trace, run `vagrant` with the verbose `-v` flag so you can see exactly where it is failing when trying to `scp` the `ssh` key.

```shell
vagrant up -debug --provider=docker
```

**Comment by rofrano** [comment]
I saw that RC3 came out today so I installed the latest RC3 and I am still having the same problem. I conducted the following experiments and I'm fairly confident it is not a networking problem:

### Experiments

1. I tried with **[X] User new virtualization framework** check and uncheck and got the same result

2. I check to see if the ports were forwarded and they are. This is the output of `docker ps`:

```
CONTAINER ID   IMAGE                             COMMAND            CREATED          STATUS          PORTS                                              NAMES
bdc2049aa9f6   rofrano/vagrant-provider:ubuntu   "/usr/sbin/init"   31 seconds ago   Up 27 seconds   127.0.0.1:5000->5000/tcp, 127.0.0.1:2222->22/tcp   lab-flask-rest_default_1617321068
```

3. I check to see if `ssh` was working and it is:

I can use `vagrant ssh` to connect to the container without a problem, so `ssh` is working with the public key. What is failing is when vagrant sees that it's using the public get and it generates a private key to be used. That is the step that failing on RC3 but works on Preview 7.

This is the output from `vagrant ssh`:

```
% vagrant ssh                                                                                                       (master)
Welcome to Ubuntu 20.04.2 LTS (GNU/Linux 5.10.25-linuxkit aarch64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

vagrant@flask:~$ 
```

### Conclusions

I’m fairly confident at this point that it’s not a network issue. I also ran vagrant with the `--debug` option and this is the relevant part of the log. I have uploaded the entire log to the Slack channel.

Relevant output from `vagrant up --debug` where it tries to remove the public key and insert the private key and `scp` fails:
```
 INFO guest: Detected: ubuntu!
DEBUG guest: Searching for cap: insert_public_key
DEBUG guest: Checking in: ubuntu
DEBUG guest: Checking in: debian
DEBUG guest: Checking in: linux
DEBUG guest: Found cap: insert_public_key in linux
DEBUG ssh: Checking whether SSH is ready...
DEBUG ssh: Re-using SSH connection.
 INFO ssh: SSH is ready!
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute:  (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 0
DEBUG guest: Searching for cap: remove_public_key
DEBUG guest: Checking in: ubuntu
DEBUG guest: Checking in: debian
DEBUG guest: Checking in: linux
DEBUG guest: Found cap: remove_public_key in linux
 INFO ssh: Inserting key to avoid password: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDbNvX7sNMt63rN986fdyupNUxg1q+FZUJNRHErZ5jSy7h+mGg6iMpt1AxYcc/E/5hxIhUiNWJI/g0i105ZUNZIKpqkJoTaY4f8BdatGzThk+2IVfR/pzZ7++wGnQPI4fBI9/B+Qe8mRo3NQhfsngFmbeSQVKrL1UQGZvB3NIyUXGF4jvfhgArEseHU8f3LSG8SGNUSMZ3bCindlpfGdPhsrDs3m8yuiROgKBkrVxkr2LErd1dQBvF1m9dC41JZgm9cl8jJVmi5Tcz5gQkwmfe4+RoAv2r/qM5kt10RJj1t0rzGotiYE3v9ut1mPjW9/asGE49UIjYfaa10qrSq8w/R vagrant
 INFO interface: detail: 
Inserting generated public key within guest...
 INFO interface: detail:     default: 
    default: Inserting generated public key within guest...
DEBUG ssh: Checking whether SSH is ready...
DEBUG ssh: Re-using SSH connection.
 INFO ssh: SSH is ready!
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute:  (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 0
DEBUG guest: Searching for cap: insert_public_key
DEBUG guest: Checking in: ubuntu
DEBUG guest: Checking in: debian
DEBUG guest: Checking in: linux
DEBUG guest: Found cap: insert_public_key in linux
 INFO guest: Execute capability: insert_public_key [#<Vagrant::Machine: default (VagrantPlugins::DockerProvider::Provider)>, "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDbNvX7sNMt63rN986fdyupNUxg1q+FZUJNRHErZ5jSy7h+mGg6iMpt1AxYcc/E/5hxIhUiNWJI/g0i105ZUNZIKpqkJoTaY4f8BdatGzThk+2IVfR/pzZ7++wGnQPI4fBI9/B+Qe8mRo3NQhfsngFmbeSQVKrL1UQGZvB3NIyUXGF4jvfhgArEseHU8f3LSG8SGNUSMZ3bCindlpfGdPhsrDs3m8yuiROgKBkrVxkr2LErd1dQBvF1m9dC41JZgm9cl8jJVmi5Tcz5gQkwmfe4+RoAv2r/qM5kt10RJj1t0rzGotiYE3v9ut1mPjW9/asGE49UIjYfaa10qrSq8w/R vagrant"] (ubuntu)
DEBUG ssh: Uploading: /var/folders/sj/r7n28_h15ps21r64tgcktw5h0000gn/T/vagrant-linux-insert-public-key20210401-2860-1bg7kg0 to /tmp/vagrant-insert-pubkey-1617321077
DEBUG ssh: Re-using SSH connection.
DEBUG ssh: Ensuring remote directory exists for destination upload
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: mkdir -p "/tmp" (sudo=false)
DEBUG ssh: stderr: 41e57d38-b4f7-4e46-9c38-13873d338b86-vagrant-ssh
DEBUG ssh: Exit status: 0
DEBUG ssh: Uploading file /var/folders/sj/r7n28_h15ps21r64tgcktw5h0000gn/T/vagrant-linux-insert-public-key20210401-2860-1bg7kg0 to remote /tmp/vagrant-insert-pubkey-1617321077
 INFO subprocess: Starting process: ["/usr/local/bin/docker", "ps", "-a", "-q", "--no-trunc"]
DEBUG subprocess: Command not in installer, not touching env vars.
 INFO subprocess: Command not in installer, restoring original environment...
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: bdc2049aa9f6e4af1d4b06d481b8afd54a1e26c8cdf7c92de86ddbb4a5ce25b6
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
 INFO subprocess: Starting process: ["/usr/local/bin/docker", "ps", "-q", "--no-trunc"]
DEBUG subprocess: Command not in installer, not touching env vars.
 INFO subprocess: Command not in installer, restoring original environment...
DEBUG subprocess: Selecting on IO
ERROR warden: Error occurred: SCP did not finish successfully (1): 
 INFO warden: Beginning recovery process...
 INFO warden: Recovery complete.
ERROR warden: Error occurred: SCP did not finish successfully (1): 
 INFO warden: Beginning recovery process...
 INFO warden: Recovery complete.
ERROR warden: Error occurred: SCP did not finish successfully (1): 
```
It repeats this over and over until I get the error output in the ticket above.

**Comment by rofrano** [comment]
I just installed Docker 3.3 on my Intel Mac and it is broken there too. Same error as on the Apple M1 Macs. 

Docker no longer works as a provider for Vagrant. This use to work until 3.3.

Does anyone know where I can download an older version of Docker? because now my primary development Mac no longer works!

**Comment by djs55** [comment]
@rofrano thanks for the detailed bug report. I see the same problem as you. Although the `--debug` logs from `vagrant` don't show the underlying error, it seems to be:

```
% scp -P 2222 -i ~/.vagrant.d/insecure_private_key Vagrantfile vagrant@localhost:/foo
Couldn't open /dev/null: Permission denied
lost connection
```

I `ssh`'d into the container and ran `chmod 666 /dev/null` (it had permissions `0440`). This allowed the provision to continue.

Normally when permissions are wrong inside a container I suspect it's a problem with the image. I'll investigate a little more but hopefully the `chmod` workaround unblocks you. Presumably it could be automated inside vagrant itself.

**Comment by rofrano** [comment]
@djs55 Thanks for those clues. I've done some more debugging and I know what's happening. It looks like Docker 3.3 has changed its behavior with using privileged containers from Docker 3.2.2. 

This is what I've found using then default `ubuntu` image from Docker Hub:

**Docker 3.3** with non-privileged container:

```sh
% docker run --rm -it -h ubuntu ubuntu bash
root@ubuntu:/# ls -l /dev/null
crw-rw-rw- 1 root root 1, 3 Apr  9 14:00 /dev/null
```

This is what I expected. The `/dev/null` privileges are `666`

**Docker 3.3** with `--privileged` flag
```sh
% docker run --rm -it -h ubuntu --privileged  ubuntu bash
root@ubuntu:/# ls -l /dev/null
crw-rw---- 1 root root 1, 3 Apr  9 14:00 /dev/null
```

Note that `/dev/null` has changed to `660`. This is what makes Vagrant fail

**Docker 3.2.2** with `--privileged` flag
```sh
% docker run --rm -it -h ubuntu --privileged  ubuntu bash
root@ubuntu:/# ls -l /dev/null
crw-rw-rw- 1 root root 1, 3 Apr  9 14:00 /dev/null
```

Note that `/dev/null` is correctly set to `666` when using Docker 3.2.2.

I don't know if this change was intentional or a defect but it broke Vagrant when using privileged containers.

**Comment by stephen-turner** [comment]
Also reported in #5559.

**Comment by djs55** [comment]
@rofrano I've got a developer build with the fix / workaround, if you'd like to try. It's Intel only (for the moment) and not notarised. The build is: https://desktop-stage.docker.com/mac/stable/amd64/62952/Docker.dmg

For me it looks better:
```
$ vagrant up --provider=docker
Bringing machine 'default' up with 'docker' provider...
==> default: Creating and configuring docker networks...
==> default: Creating the container...
    default:   Name: lab-flask-rest_default_1617991989
    default:  Image: rofrano/vagrant-provider:ubuntu
    default: Volume: /Users/djs/bugs/vagrant/lab-flask-rest:/vagrant
    default:   Port: 127.0.0.1:5000:5000
    default:   Port: 127.0.0.1:2222:22
    default:  
    default: Container created: 367969c6eb800bef
==> default: Enabling network interfaces...
==> default: Starting container...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default: 
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Running provisioner: file...
    default: ~/.gitconfig => ~/.gitconfig
==> default: Running provisioner: file...
    default: ~/.ssh/id_rsa => ~/.ssh/id_rsa
==> default: Running provisioner: file...
    default: ~/.ssh/id_rsa.pub => ~/.ssh/id_rsa.pub
==> default: Running provisioner: shell...
    default: Running: inline script
    default: Get:1 http://archive.ubunt
...
```

**Comment by rofrano** [comment]
@djs55 YES! It works like a charm for me too on Intel. I can't want to get an Apple Silicon build. Thank you so much!

**Comment by djs55** [comment]
@rofrano thanks for the confirmation! If you'd like to try an Apple Silicon build, this one has the fix in: https://desktop-stage.docker.com/mac/stable/arm64/62981/Docker.dmg

**Comment by rofrano** [comment]
@djs55 You are 2 for 2! This build works as expected on Apple Silicon. Thank you so much for fixing the issue as quickly as you did. I really appreciate it. 

I may not have mentioned this, but I teach a graduate course on DevOps at NYU and I use Vagrant and VirtualBox as our "infrastructure as code" development environment for the hands-on labs. This semester, 8 of my students showed up with Apple M1 Macs and if it weren't for Docker on Apple Silicon, and the the ability to use Docker as a provider for Vagrant, I wouldn't have had any way to produce automated virtual environments for these students. So my students thank you too.

You Rock! 👍

**Comment by djs55** [comment]
@rofrano thanks for letting us know and also for the very clear bug report. The fix was definitely a team effort!

It's nice to hear how Docker is being used -- it sounds like a fun course.

I'll leave this issue open until the fix is shipped in the upcoming 3.3.1 release, hopefully next week.

Enjoy the rest of your weekend.

**Comment by davidkarlsen** [comment]
@djs55 this also affects mac on intel. Do you have any fresh builds for intel?

**Comment by toph-allen** [comment]
Similar to to [docker/for-win#10910](https://github.com/docker/for-win/issues/10910) this bug affecting `/dev/random` and `/dev/urandom` triggered [this bug](https://stackoverflow.com/questions/61381288/linux-apache-python-cgi-error-failed-to-get-random-numbers-to-initialize-python) and caused builds to fail.

**Comment by stephen-turner** [comment]
Thanks all. We've discovered that this was caused by a change in umask behaviour in one of our upstream components, runc. We're working round the change in behaviour.

**Comment by x8k** [comment]
> @rofrano I've got a developer build with the fix / workaround, if you'd like to try. It's Intel only (for the moment) and not notarised. The build is: https://desktop-stage.docker.com/mac/stable/amd64/62952/Docker.dmg

I couldn't resist and I tried it, it works like a charm on Mac Intel&Catalina :smiling_face_with_three_hearts:

**Comment by mat007** [comment]
The fix has now been released with Docker Desktop 3.3.1. I’ll go ahead and close this issue. Thanks!
